### ISSUE TITLE
torch.compile produces wrong gradient for diagonal_scatter backward

### ISSUE THREAD

**Issue by ALinrunrun** [pytorch/pytorch#183602]
### 🐛 Describe the bug

`torch.compile` produces an incorrect gradient for `torch.diagonal_scatter` in backward.

Eager execution gives the expected non-zero gradient for `x`, while the compiled version produces a zero gradient.

```
import torch

def fn(x, src):
    return torch.diagonal_scatter(x, src, 0).sum()

x1 = torch.zeros(5, 5, requires_grad=True)
s1 = torch.ones(5, requires_grad=True)
fn(x1, s1).backward()

torch._dynamo.reset()

x2 = torch.zeros(5, 5, requires_grad=True)
s2 = torch.ones(5, requires_grad=True)
torch.compile(fn)(x2, s2).backward()

print("eager x.grad sum:", float(x1.grad.sum()))
print("compile x.grad sum:", float(x2.grad.sum()))
```

### Error logs

```
eager x.grad sum: 20.0
compile x.grad sum: 0.0
```

### Versions

PyTorch version: 2.11.0+cu130
Is debug build: False
CUDA used to build PyTorch: 13.0
ROCM used to build PyTorch: N/A

OS: Ubuntu 24.04.2 LTS (x86_64)
GCC version: (Ubuntu 13.3.0-6ubuntu2~24.04.1) 13.3.0
Clang version: Could not collect
CMake version: Could not collect
Libc version: glibc-2.39

Python version: 3.10.20 (main, Mar 11 2026, 17:46:40) [GCC 14.3.0] (64-bit runtime)
Python platform: Linux-5.15.167.4-microsoft-standard-WSL2-x86_64-with-glibc2.39
Is CUDA available: False
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA GeForce RTX 3080 Laptop GPU
Nvidia driver version: 545.92
cuDNN version: Could not collect
Is XPU available: False
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True
Caching allocator config: N/A

Versions of relevant libraries:
[pip3] numpy==2.2.6
[pip3] onnx==1.21.0
[pip3] onnx2torch==1.5.15
[pip3] onnxruntime==1.23.2
[pip3] torch==2.11.0
[pip3] torchvision==0.26.0
[pip3] triton==3.6.0

cc @pragupta @chauhang @penguinwu

**Comment by khushi-411** [comment]
Hi, I'm working on this. Will share a PR soon :)
